pkg/types/azure/validation/machinepool.go

```diff
@@ -12,5 +19,66 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
 	if p.OSDisk.DiskSizeGB < 0 {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("diskSizeGB"), p.OSDisk.DiskSizeGB, "Storage DiskSizeGB must be positive"))
 	}
+
+	return allErrs
+}
+
+// ValidateDiskType checks that the specified disk type is valid for control plane.
+func ValidateDiskType(p *types.MachinePool, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if p.Name == "master" && p.Platform.Azure.OSDisk.DiskType == "Standard_LRS" {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("diskType"), p.Platform.Azure.OSDisk.DiskType, fmt.Sprintf("%s not compatible with control planes.", p.Platform.Azure.OSDisk.DiskType)))
+	}
+
+	if p.Platform.Azure.OSDisk.DiskType != "" {
+		diskTypes := sets.NewString("Standard_LRS", "StandardSSD_LRS", "Premium_LRS") // "UltraSSD_LRS"
+
+		if !diskTypes.Has(p.Platform.Azure.OSDisk.DiskType) {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("diskType"), p.Platform.Azure.OSDisk.DiskType, diskTypes.List()))
+		}
+	}
+
+	return allErrs
+}
+
+// ValidateDefaultDiskType checks that the specified disk type is valid for default Azure Machine Platform.
+func ValidateDefaultDiskType(p *azure.MachinePool, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if p.OSDisk.DiskType != "" {
+		diskTypes := sets.NewString("StandardSSD_LRS", "Premium_LRS") // "UltraSSD_LRS"
+
+		if !diskTypes.Has(p.OSDisk.DiskType) {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("diskType"), p.OSDisk.DiskType, diskTypes.List()))
+		}
+	}
+
+	// if p.OSDisk.DiskType == "UltraSSD_LRS" {
+	// 	allErrs = append(allErrs, ValidateRegionForUltraDisks(fldPath, p)...)
+	// }
+
+	return allErrs
+}
+
+// ValidateRegionForUltraDisks checks that the Ultra SSD disks are available for the user.
+func ValidateRegionForUltraDisks(fldPath *field.Path, region string) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer cancel()
+	client, err := azconfig.NewClient(ctx)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
+	}
+
+	contxt, canc := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer canc()
+
+	_, err = client.GetResourceSku(contxt, region, "UltraSSD_LRS")
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
+	}
+
```

Comment:
this function depends on pkg/asset, please move it there. pkg/types shuoldn't depend on that package,